### PR TITLE
Bulk add performance improvements

### PR DESCRIPTION
**Pull request by symroe**
Second branch today that is badly named!

See each commit for info. This is trying to reduce the number of DB queries used per add and update operation in the bulk adding app.

### CODE REVIEW

File: ynr/apps/bulk_adding/helpers.py [hunk 1]
```diff
@@ -4,21 +4,19 @@
                                raise_if_unsafe_to_delete)
 from candidates.models.auth import check_creation_allowed
 from candidates.views.version_data import get_change_metadata, get_client_ip
+from tasks.models import connect_task_signal, disconnect_task_signal
 
 
 def add_person(request, person_data):
+    disconnect_task_signal()
```
Comment (chris48s) [inline ynr/apps/bulk_adding/helpers.py]: Is there any risk here that we might disconnect at the top of this function, throw an exception somewhere in the body and then not re-connect at the end, leaving us in an inconsistent state? Would it be better to

```python
disconnect_task_signal()
try:
    # do stuff
finally:
    connect_task_signal()
```

File: ynr/apps/bulk_adding/tests/test_bulk_add.py [hunk 2]
```diff
@@ -88,7 +88,8 @@ def test_submitting_form(self):
         response = response.follow()
         form = response.forms[1]
         form['form-0-select_person'].select('_new')
-        response = form.submit()
+        with self.assertNumQueries(70):
```
Comment (chris48s) [inline ynr/apps/bulk_adding/tests/test_bulk_add.py]: +1 for `assertNumQueries()` - making assertions about application performance is really useful.

Simultaneously, this seems like a bit of an arbitrary number. As a reader of this file in isolation, I have no idea why this is 70 (or 66 in the other test). I think these tests could benefit from some sort of explanation of these numbers.

It also seems like this is a test that is going to break quite often. I'm not sure I have a great suggestion for improvement here. One thing I have done in https://github.com/DemocracyClub/uk-geo-utils/blob/master/uk_geo_utils/tests/test_addressbase_geocoder.py is express the performance requirements in terms of a range. It works there because the range is meaningful though. I just feel like this is a number we're often going to find ourselves arbitrarily incrementing/decrementing as we modify code in tangentally related places.. Maybe performance assertions would be best reserved for the more granular unit test in `ynr/apps/bulk_adding/tests/test_helpers.py` where there are less moving parts rather than a high level integration test like this. Do you have any thoughts?

Also, am I reading this correctly that it takes 70 round trips to the database to insert one candidate, and _that's an improvement_ ? That poor database! :( Presumably (looking at the other tests) that's 61 queries of setup/bootstrapping that must occur regardless of whether we are inserting 1 candidate or 50 and then 9 queries per candidate inserted? That's the kind of info that would at least be useful in a comment here.

**Comment by chris48s** [comment]
This probably isn't something that should hold up the merge here, but maybe another idea to think about if we need to optimise this more in future is whether there is scope to use `bulk_create()`.

I know that there is a lot of FKs and hooks/events which makes things complex here, but I wonder if there would be scope to (for example), have `add_person()` create the `Person` and `PersonExtra` records and then return a new `LoggedAction` object (which hasn't been saved yet) and persist all the log entries into the DB in one `bulk_insert`. As long as the whole lot is in a transaction, that would be fine.

Something to think about..

**Comment by symroe** [comment]
So the reason this branch is called `batch_update_in_bulk_add` is because that was my first idea…

I agree it would be better, however there are a lot of things that prevent that from happening, and I wasn't going to start gutting everything this soon before an election – something for future.

To give some examples: 

1. The `Base/Extra` pairing means that it's already a little tricky to save anything. I tried adding all the base models to a list, then all the extra ones. Bulk add the first, update the second with the base pk and then save the second. This works ok, but it's a bit messy.

2. Then I realised that this isn't really where the bulk of the queries come from. That's in the `record_version` and `old_memberships` sections. Due to the way `record_version` works, I had a choice: either don't record a version for the models created from bulk adding or call `record_version` after each model was bulk created. This would have ended up saving 2 INSERTs per person, leaving ~25 queries per person, but with quite hard to follow code.

3. Then I looked in to the queries, and it's basically caused by using generic foreign keys in a lot of places, and generally having an abstracted data model. `record_version`, `get_change_metadata` and [`get_person_as_version_data`](https://github.com/DemocracyClub/yournextrepresentative/blob/08acbade5b2f1b483b452a15946485c3edb7128d/ynr/apps/candidates/models/versions.py#L16) basically build a JSON representation of the model and all related models. This is an expensive thing to do when the [data model is a bit ad-hoc and stored in the database itself](https://github.com/DemocracyClub/yournextrepresentative/blob/08acbade5b2f1b483b452a15946485c3edb7128d/ynr/apps/candidates/models/popolo_extra.py#L360).

There is room for some improvements in future, like caching / hard coding `SimplePopoloField.objects.all()`, `ComplexPopoloField.objects.all()` and `ExtraField.objects.all()` (we've changed their return values once in the last couple of years).

Also, I think the number of queries would come down (not quite by half) when we merge all the base/extra models in to a single model – Membership is the top of my list at the moment.

All in all…it's a little messy and changing too much risks breaking other parts of the site that we really don't want to be breaking (like version history, merging, etc).

**Comment by chris48s** [comment]
All of those are useful points. I'd forgotten that we rely really heavily on 1-to-1 models in this project which is a performance killer because it basically doubles the number of DB round-trips on a lot of operations.

If nothing else, it is useful to get all that info written down somewhere for future reference.
